chore: migrate to node action

Jexsie opened the pull request:
**Description**:
<!--
One or two line summary of what this PR does and why it is needed, followed by a list
of changes in imperative, present tense for use in the commit message or changelog. Example:


* Add config property
* Change column name
* Remove ...
-->
This PR introduces a complete rewrite of the Hiero Solo GitHub Action from a composite  action to a Node.js-based action. It incorporates all improvements, debugging, and design decisions from our recent implementation discussions

**Related issue(s)**:

Fixes https://github.com/hiero-ledger/hiero-solo-action/issues/81

**Notes for reviewer**:
<!-- Provide logs, performance numbers or screenshots of the new functionality -->

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

Jexsie committed:
migrate to node actions

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
rewrite solo deployment and account creation in Node.js

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
add install and build steps

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
execute failing commands in bash

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
used child processes

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
add missed step

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
return outputs correctly

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
add missed port forward

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix port forwarding

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
add comments for each step

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
revert fix

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
disable all manual port forwardings

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
remove unused imports

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
add debuger step

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
improve port forwarding

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
more fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

hendrikebbers commented on .github/workflows/validation.yml:
```diff
@@ -9,6 +9,35 @@ jobs:
   validate-hedera-outputs:
     runs-on: ubuntu-latest
     steps:
+      - name: Setup Java
```
This is wrong. All steps done here must be part of the action we provide. This file is just to check / test the action and should only call our action. You can see this file as "This is how an external repo would use it". We do not want external repos to install java / kind /... by hand. With that we need to figure out how we can get those steps now in the new / refactored action

Jexsie replied:
Resolved using https://github.com/marketplace/actions/tool-cache

Jexsie committed:
Use @vercel/ncc for building files
Remove the python account information extractor file
Update import to fix code analysis in the post.ts file

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Ndacyayisenga-droid commented on .github/workflows/validation.yml:
```diff
@@ -9,6 +9,35 @@ jobs:
+
+      - name: Install Solo
+        shell: bash
+        run: npm install -g @hashgraph/solo@0.39.0
```
Use v0.40.1

jeromy-cannon replied:
now at v0.69.0.

Jexsie committed:
create a setup-env action

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix build and install step

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
remove dist from gitignore

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
use the bundled post.js

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
improve type handling

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
improve error handling

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
improve error handling in post.ts

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
etup husky to auto-commit dist on changes

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

jeromy-cannon commented on action.yml:
```diff
@@ -35,402 +43,29 @@ inputs:
-  icon: "share-2"
-  color: "black"
+runs:
+  using: node20
```
node 20 is no longer supported

Jexsie committed:
chore: use pnpm as the package mananger

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
chore: upgrade to node 24

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
feat: migrate system dependency installation from composite action to TypeScript implementation

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
remove setup action

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
feat: update system dependency installation to include Java 21, Kind v0.26.0, kubectl v1.31.4, and Solo CLI

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
refactor: remove redundant setup-env action from validation workflow jobs

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
refactor: move utility functions to separate module and update tsconfig settings

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
chore: upgrade Kind and Kubectl versions, replace info with safeInfo, and add --dev flag to solo init

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
feat: add Solo CLI version detection and support for v0.44.0+ command syntax

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
chore: remove redundant fs dependency and use the naive node:fs

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
chore: remove redundant fs dependency and use the naive node:fs

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
chore: update dependencies to latest versions

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
chore: update tsconfig for ESM compatibility, fix import extensions

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
chore: update tsconfig for ESM compatibility, fix import extensions

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
update node

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
feat: simply solo version capture logic

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
refactor: replace direct exec calls with safeExec and move isVersionGte to utils

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
feat: replace apt-based dependency installation with tool-cache downloads for Python, wget, Java, and kubectl

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
refactor: modularize dependency installation and constants into separate files

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
refactor: replace safeExec with runCommand and soloRun for streamlined command execution

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
refactor: use safe commands in the post script

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
chore: add ESLint and Prettier configuration with updated project dependencies

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
chore: add lint changes

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
test node action on self-hosted runners

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
add a check for node on selfhosted runners

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
feat: add cleanup for Solo config directory and prune Docker resources to ensure fresh state on self-hosted runners

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
feat: add support for installing jq in setup action

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
chore: add concurrency group to validation workflow to cancel in-progress runs

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
refactor: centralize infrastructure constants and integrate linting into pre-commit workflow

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
refactor: move safeSetFailed utility from index.ts to utils.ts for better modularity

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
feat: implement tool-cache lookup to prevent redundant downloads for all dependencies

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
switch back to ubuntu latest

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
refactor: improve type safety in catch blocks and replace logical OR with nullish coalescing operator

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix: update eslint linting command to auto-fix issues

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
feat: add support for optional block node deployment before network initialization

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
refactor: fix consensus network deployment and mirror node service management with port forwarding support

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
recover extract account python file since it is used for the local setup

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
refactor: remove redundant safe-wrapper utility functions in favor of direct standard library calls

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>

Jexsie committed:
fix up

Signed-off-by: Jessie Ssebuliba <jessiessebuliba@gmail.com>
